src/persist-client/src/operators/shard_source.rs

```diff
@@ -298,75 +216,192 @@ where
             // subscription and pass them on.
-            let mut batch_parts = vec![];
             while inflight_bytes < max_inflight_bytes {
-                match pinned_stream.next().await {
-                    Some(Ok(ListenEvent::Updates(mut parts))) => {
-                        batch_parts.append(&mut parts);
+                tokio::select! {
+                    // If the token has been dropped, we begin a graceful shutdown
+                    // by downgrading to the empty frontier and proceeding to the
+                    // next state, where we await our feedback edge's progression.
+                    //
+                    // NB: Reading from a channel is cancel safe.
+                    _ = token_tx.closed() => {
+                        cap_set.downgrade(&[]);
+                        break 'emitting_parts;
                     }
-                    Some(Ok(ListenEvent::Progress(progress))) => {
-                        let session_cap = cap_set.delayed(&current_ts);
-                        let mut descs_output = descs_output.activate();
-                        let mut descs_session = descs_output.session(&session_cap);
-
-                        // NB: in order to play nice with downstream operators whose invariants
-                        // depend on seeing the full contents of an individual batch, we must
-                        // atomically emit all parts here (e.g. no awaits).
-                        let bytes_emitted = {
-                            let mut bytes_emitted = 0;
-                            for part_desc in std::mem::take(&mut batch_parts) {
-                                bytes_emitted += part_desc.encoded_size_bytes();
-                                // Give the part to a random worker. This isn't
-                                // round robin in an attempt to avoid skew issues:
-                                // if your parts alternate size large, small, then
-                                // you'll end up only using half of your workers.
-                                //
-                                // There's certainly some other things we could be
-                                // doing instead here, but this has seemed to work
-                                // okay so far. Continue to revisit as necessary.
-                                let worker_idx = usize::cast_from(Instant::now().hashed()) % num_workers;
-                                descs_session.give((worker_idx, part_desc.into_exchangeable_part()));
+                    // Return the leases of any parts that the following stage,
+                    // `shard_source_fetch` has finished reading. This allows us
+                    // to advance our SeqNo hold as we continue to emit batches.
+                    //
+                    // NB: AsyncInputHandle::next is cancel safe
+                    Some(completed_fetch) = completed_fetches.next() => {
```

Comment:
Won't this structure make it so that we only return leases when `inflight_byte < max_inflight_bytes`? I don't have a good solution, though. 🙈

Reply:
I'll play with it and see. my initial read was that it's not super problematic if we only return parts when flow control isn't being applied, but I bet we can just squeeze another `select` in over `completed_fetches` and `flow_control_input`

Reply:
updated! I moved the flow control case into the `select!` and wrapped it and the normal path behind `if` preconditions. assuming I'm understanding all the implications correctly, I actually think this reads more easily than before